src/http/client.cc

```diff
@@ -19,10 +19,79 @@
  * Copyright (C) 2022 Scylladb, Ltd.
  */
 
+#include <seastar/core/loop.hh>
+#include <seastar/core/when_all.hh>
 #include <seastar/http/client.hh>
+#include <seastar/http/request.hh>
+#include <seastar/http/reply.hh>
+#include <seastar/http/response_parser.hh>
 
 namespace seastar {
 namespace http {
 
+client::client(connected_socket&& fd)
+        : _fd(std::move(fd))
+        , _read_buf(_fd.input())
+        , _write_buf(_fd.output())
+{
+}
+
+future<> client::write_body(request& req) {
+    if (req.body_writer) {
+        return req.body_writer(_write_buf);
+    } else {
+        return make_ready_future<>();
+    }
+}
+
+future<> client::send_request(request& req) {
+    if (req._version.empty()) {
+        req._version = "1.1";
+    }
+    if (req.content_length != 0) {
+        if (!req.body_writer) {
```

Comment:
I think a zero length body should be allowed... So the marker for using the body_writer should be something else.

Reply:
You mean that it should be valid to send a request with "Content-Length" non zero, but with empty body? How is this supposed to be used?

Reply:
No, I mean it is valid to send content-length==0 and an empty body string. Why does the fact you set content-length==0 mean you wanted to use a body_writer function?

Reply:
But the check is opposite -- if the content_length is non zero, then the caller must provide the body_writer. It's sort of a sanity check. Otherwise the request will be sent, but the server will hang waiting for the body to arrive.

Reply:
Ok, I got mixed up on that !=0, but ll don't understand: Why is it not possible to have content_length != 0, and then provide the request via an already-prepared string, and NOT via a body writer?
Isn't it actually supposed to be the other way around? If content_length != 0, you shouldn't use a body writer (which can write an unknown amount of data)?
If you use a body writer, shouldn't you use chunked encoding to write the unknown amount of data? This is what we do in for the response.

Reply:
> Why is it not possible to have content_length != 0, and then provide the request via an already-prepared string, and NOT via a body writer?

Because currently there's no way to provide a request body other than via a body_writer. 

> If you use a body writer, shouldn't you use chunked encoding to write the unknown amount of data?

That's the same -- if you give user a stream to write body (or its chunks) into, its size anyway goes _before_ the data, so there's still a chance someone will mis-use this and write more or less data into the stream than it was supposed to

> This is what we do in for the response.

Mm, not quite. The `https::reply` doesn't have the `content_length` member at all. And since the `httpd::reply::_headers` is public, and `httpd::reply::write_body()` doesn't make any assumptions about what user may write, there's still an option to write wrong amount of data. The only "safe" (in this sense) way to send the body is by providing a prepared string, as you suggested above. I will add this option in the next version.

Reply:
> > Why is it not possible to have content_length != 0, and then provide the request via an already-prepared string, and NOT via a body writer?
> 
> Because currently there's no way to provide a request body other than via a body_writer.

Oh, I didn't realize this. This is not the case for the HTTP server. So will you have the simpler single-string option as a FIXME?

> > If you use a body writer, shouldn't you use chunked encoding to write the unknown amount of data?
> 
> That's the same -- if you give user a stream to write body (or its chunks) into, its size anyway goes _before_ the data, so there's still a chance someone will mis-use this and write more or less data into the stream than it was supposed to

But you are missing the fact the that whole raison-d'etre of why chunked encoding was added to HTTP 1.1 was to allow the sender to **not need to calculate content-length in advance**. You can always prepare the next chunk (e.g., 1KB) and send it, without needing to calculate in advance that the entire response (or in your code, request) will have 1,234,567 bytes.

So in HTTP 1.1, when you use chunked encoding, you **don't** specify a content-length. The RFC explicitly says:
> Messages MUST NOT include both a Content-Length header field and a non-identity transfer-coding.
This is why I find it surprising or even wrong that you support this combination as the only option.

Reply:
> Oh, I didn't realize this. This is not the case for the HTTP server. So will you have the simpler single-string option as a FIXME?

Yup. Will add to TODO.

> But you are missing the fact the that whole raison-d'etre of why chunked encoding was added to HTTP 1.1 was to allow the sender to not need to calculate content-length in advance. You can always prepare the next chunk (e.g., 1KB) and send it, without needing to calculate in advance that the entire response (or in your code, request) will have 1,234,567 bytes.
> 
> So in HTTP 1.1, when you use chunked encoding, you don't specify a content-length. The RFC explicitly says:

You're perfectly right, but my point is about another thing. If we want to ensure that length and contents (or whatever -- Content-Length body of chunks of the chunked body) perfectly match each other, then giving the caller raw output stream is not the way to go. User may write length of 10 and then feed 20 bytes into the stream, nothing prevents him from doing it. And vice-verse, user may write the length of 20 and then put 10 bytes into the stream, current output_stream API doesn't have any calls to check this. So once there's a body-writer lambda caller has access to, then it's up to him to provide correct length + content pair.